### PR TITLE
Implement MSC3912: Relation-based redactions

### PR DESCRIPTION
**Pull request by babolivier**
Implements MSC3912

Part of #14231 

Can be reviewed commit by commit.

This implementation is lacking a way to keep watch for incoming events that match the redaction criteria after the redaction request finishes, that'll come in a separate PR.

### CODE REVIEW

File: synapse/storage/databases/main/relations.py [hunk 1]
```diff
@@ -295,6 +295,44 @@ def _get_recent_references_for_event_txn(
...
+        """
+
+        def get_all_relation_ids_for_event_with_types_txn(
+            txn: LoggingTransaction,
+        ) -> List[str]:
+            sql = """
+            SELECT event_id FROM event_relations
+            WHERE
+                relates_to_id = ?
+                AND relation_type IN (%s)
```
Comment (erikjohnston) [inline synapse/storage/databases/main/relations.py:322]: I think we can use `simple_select_many` here?

File: synapse/rest/client/room.py [hunk 2]
```diff
@@ -1045,20 +1047,45 @@ async def on_POST(
...
+                        "content": content,
+                        "room_id": room_id,
+                        "sender": requester.user.to_string(),
+                        "redacts": event_id,
+                    },
+                    txn_id=txn_id,
+                )
+
+                if with_relations:
+                    run_in_background(
```
[the remark this answers is not in the document]
Comment (babolivier) [reply synapse/rest/client/room.py:1081]: Indeed. I keep confusing the two :(

**Comment by babolivier** [comment]
Adding back to the synapse-core queue since Erik is away for the rest of the week.

**Review by squahtx — APPROVED** [review]
lgtm!

**Pull request merged**